consensus/cometbft/service/finalize_block.go

```diff
@@ -94,9 +98,66 @@ func (s *Service) finalizeBlockInternal(
 		TxResults:             txResults,
 		ValidatorUpdates:      valUpdates,
 		ConsensusParamUpdates: &cp,
+		NextBlockDelay:        s.nextBlockDelay(req),
 	}, nil
 }
 
+// Panics if SBT upgrade parameters are not set.
+//
+//nolint:lll // long message on one line for readability.
+func (s *Service) nextBlockDelay(req *cmtabci.FinalizeBlockRequest) time.Duration {
+	delay := constBlockDelay
+
+	switch {
+	case s.blockDelay == nil: // height > 0; blockDelay record doesn't exist in DB (first live upgrade)
+		if s.sbtUpgradeHeight == 0 {
+			panic("SBT (stable block time) upgrade detected. --beacon-kit.cometbft.sbt-upgrade-height (block's height where SBT was enabled) must be set")
+		}
+		switch {
+		case req.Height > s.sbtUpgradeHeight: // upgrade happened in the past
+			if s.sbtUpgradeTime.IsZero() {
+				panic("SBT (stable block time) upgrade detected. --beacon-kit.cometbft.sbt-upgrade-time (block's time where SBT was enabled) must be set")
+			}
+			s.blockDelay = BlockDelayUponGenesis(
+				s.sbtUpgradeTime,
+				s.sbtUpgradeHeight,
+			)
+			delay = s.blockDelay.Next(req.Time, req.Height)
+		case req.Height == s.sbtUpgradeHeight: // upgrade is happening now
+			s.blockDelay = BlockDelayUponGenesis(
+				req.Time,
```

Comment:
Also wondering if, operationally, it would be smarted here to use `s.sbtUpgradeTime` instead of `req.Time`. Note that in this scenario `req.Height == s.sbtUpgradeHeight`